blocks/library/block/index.js

```diff
@@ -0,0 +1,172 @@
+import ReusableBlockEditPanel from './edit-panel';
+
+class ReusableBlockEdit extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.startEditing = this.startEditing.bind( this );
+		this.stopEditing = this.stopEditing.bind( this );
+		this.setAttributes = this.setAttributes.bind( this );
+		this.setName = this.setName.bind( this );
+		this.updateReusableBlock = this.updateReusableBlock.bind( this );
+
+		this.state = {
+			isEditing: false,
+			name: null,
+			attributes: null,
+		};
+	}
+
+	componentDidMount() {
+		if ( ! this.props.reusableBlock ) {
+			this.props.fetchReusableBlock();
+		}
+	}
+
+	startEditing() {
+		this.setState( { isEditing: true } );
+	}
+
+	stopEditing() {
+		this.setState( {
+			isEditing: false,
+			name: null,
+			attributes: null,
+		} );
+	}
+
+	setAttributes( attributes ) {
+		this.setState( ( prevState ) => ( {
+			attributes: { ...prevState.attributes, ...attributes },
+		} ) );
+	}
+
+	setName( name ) {
+		this.setState( { name } );
+	}
+
+	updateReusableBlock() {
+		const { name, attributes } = this.state;
+
+		// Use pickBy to include only changed (assigned) values in payload
+		const payload = pickBy( {
+			name,
+			attributes,
+		} );
+
+		this.props.updateReusableBlock( payload );
+		this.props.saveReusableBlock();
+		this.stopEditing();
+	}
+
+	render() {
+		const { focus, reusableBlock, convertBlockToStatic } = this.props;
+		const { isEditing, name, attributes } = this.state;
+
+		if ( ! reusableBlock ) {
+			return <Placeholder><Spinner /></Placeholder>;
+		}
+
+		const reusableBlockAttributes = { ...reusableBlock.attributes, ...attributes };
+		const blockType = getBlockType( reusableBlock.type );
+		const BlockEdit = blockType.edit || blockType.save;
+
+		// Generate a class name for the block's editable form
+		const generatedClassName = hasBlockSupport( blockType, 'className', true ) ?
+			getBlockDefaultClassname( reusableBlock.type ) :
+			null;
+		const className = classnames( generatedClassName, reusableBlockAttributes.className );
+
+		return [
+			// We fake the block being read-only by wrapping it with an element that has pointer-events: none
+			<div key="edit" style={ { pointerEvents: isEditing ? 'auto' : 'none' } }>
+				<BlockEdit
+					{ ...this.props }
+					focus={ isEditing ? focus : null }
+					attributes={ reusableBlockAttributes }
+					setAttributes={ isEditing ? this.setAttributes : noop }
+					className={ className }
+				/>
+			</div>,
+			focus && (
+				<ReusableBlockEditPanel
+					key="panel"
+					isEditing={ isEditing }
+					name={ name !== null ? name : reusableBlock.name }
+					isSaving={ reusableBlock.isSaving }
+					onEdit={ this.startEditing }
+					onDetach={ convertBlockToStatic }
+					onChangeName={ this.setName }
+					onSave={ this.updateReusableBlock }
+					onCancel={ this.stopEditing }
+				/>
+			),
+		];
+	}
+}
+
+const ConnectedReusableBlockEdit = connect(
```

Comment:
I don't have any ideas at the moment and it may have already been mentioned, but we need to work to find a solution where the block can access state without the implicit dependency on editor, either:

- Blocks module having its own separate state
- Merging more of [the non-post-editor-specific-] editor and blocks

Obvious too by the fact that we have equivalent action creators for what we're dispatching here, but unable (unwilling) to import them directly from editor.

Reply:
Agreed, and I also have no ideas on how to address this right now. 

> Merging more of [the non-post-editor-specific-] editor and blocks

Could you elaborate on this? I'm not quite caught up on my history here. Did we abandon the effort to merge the two modules (https://github.com/WordPress/gutenberg/pull/2795)?

Reply:
Just an idea: Does it make more sense for this block to be defined in the `editor` module instead?

I also even wonder if the `blocks` module should hold only the blocks API and define all the blocks in the `editor` module

Reply:
> Just an idea: Does it make more sense for this block to be defined in the `editor` module instead?

I think there's a lot of merit to that idea.

> I also even wonder if the `blocks` module should hold only the blocks API and define all the blocks in the `editor` module

That one I'm not too convinced of right now. I mean, I could picture using `editor/library`, but the current situation of having the library sit next to the API makes sense to me, as the core blocks are the canonical blocks: in that sense, they highly complement the API (both code and docs) by providing numerous well tested use cases.